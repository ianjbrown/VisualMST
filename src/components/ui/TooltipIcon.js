import FirstPageIcon from "@mui/icons-material/FirstPage";
import ArrowBackIosNewOutlinedIcon from "@mui/icons-material/ArrowBackIosNewOutlined";
import ArrowForwardIosOutlinedIcon from "@mui/icons-material/ArrowForwardIosOutlined";
import LastPageIcon from "@mui/icons-material/LastPage";
import PlayButtonToggle from "./PlayButtonToggle";
import HideGraphToggle from "./HideGraphToggle";
import classes from "./PlayBack.module.css";
import { OverlayTrigger, Tooltip } from "react-bootstrap";

function TooltipIcon(props) {
  function firstHandler() {
    props.firstHandler();
  }

  function backHandler() {
    props.backHandler();
  }

  function pauseHandler() {
    props.pauseHandler();
  }

  function forwardHandler() {
    props.forwardHandler();
  }

  function lastHandler() {
    props.lastHandler();
  }

  function toggleHandler() {
    props.toggleHandler();
  }

  switch (props.id) {
    case "first":
      return (
        <OverlayTrigger
          placement={"top"}
          overlay={
            <Tooltip id={`tooltip-top`}>
              <strong>First Step</strong>
            </Tooltip>
          }
        >
          <FirstPageIcon
            sx={{ "&:hover": { cursor: "pointer" }, fontSize: 60 }}
            onClick={firstHandler}
          />
        </OverlayTrigger>
      );
    case "back":
      return (
        <OverlayTrigger
          placement={"top"}
          overlay={
            <Tooltip id={`tooltip-top`}>
              <strong>Previous Step</strong>
            </Tooltip>
          }
        >
          <ArrowBackIosNewOutlinedIcon
            sx={{ "&:hover": { cursor: "pointer" }, fontSize: 60 }}
            onClick={backHandler}
          />
        </OverlayTrigger>
      );
    case "play":
      return (
        <OverlayTrigger
          placement={"top"}
          overlay={
            <Tooltip id={`tooltip-top`}>
              <strong>
                {props.paused ? "Play Visualisation" : "Pause Visualisation"}
              </strong>
            </Tooltip>
          }
        >
          <div>
            <PlayButtonToggle paused={props.paused} onClick={pauseHandler} />
          </div>
        </OverlayTrigger>
      );
    case "forward":
      return (
        <OverlayTrigger
          placement={"top"}
          overlay={
            <Tooltip id={`tooltip-top`}>
              <strong>Next Step</strong>
            </Tooltip>
          }
        >
          <ArrowForwardIosOutlinedIcon
            sx={{ "&:hover": { cursor: "pointer" }, fontSize: 60 }}
            onClick={forwardHandler}
          />
        </OverlayTrigger>
      );
    case "last":
      return (
        <OverlayTrigger
          placement={"top"}
          overlay={
            <Tooltip id={`tooltip-top`}>
              <strong>Final Step</strong>
            </Tooltip>
          }
        >
          <LastPageIcon
            sx={{ "&:hover": { cursor: "pointer" }, fontSize: 60 }}
            onClick={lastHandler}
          />
        </OverlayTrigger>
      );
    case "hide":
      return (
        <OverlayTrigger
          placement={"top"}
          overlay={
            <Tooltip id={`tooltip-top`}>
              <strong>
                {props.backgroundOn
                  ? "Hide original graph edges"
                  : "Show original graph edges"}
              </strong>
            </Tooltip>
          }
        >
          <div className={classes.barToggle}>
            <HideGraphToggle
              backgroundOn={props.backgroundOn}
              toggleHandler={toggleHandler}
            />
          </div>
        </OverlayTrigger>
      );
  }
}

export default TooltipIcon;
