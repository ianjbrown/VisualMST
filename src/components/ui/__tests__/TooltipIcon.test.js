import TooltipIcon from "../TooltipIcon";
import {
  render,
  screen,
  cleanup,
  waitFor,
} from "@testing-library/react";
import "@testing-library/jest-dom/";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";

afterEach(() => {
  cleanup();
});

const buttons = [
  { symbol: "FirstPageIcon", text: "First Step", id: "first" },
  { symbol: "ArrowBackIosNewOutlinedIcon", text: "Previous Step", id: "back" },
  { symbol: "ArrowForwardIosOutlinedIcon", text: "Next Step", id: "forward" },
  { symbol: "LastPageIcon", text: "Final Step", id: "last" },
].map(button => Object.assign(button, { toString: function() { return this.id; }}));

test.each(buttons)(`%s matches snapshot`, (button) => {
  const toolTip = renderer.create(<TooltipIcon id={button.id} />).toJSON();
  expect(toolTip).toMatchSnapshot();
});

test("play matches snapshot", () => {
  const toolTip = renderer.create(<TooltipIcon id={"play"} />).toJSON();
  expect(toolTip).toMatchSnapshot();
});

test("hide matches snapshot", () => {
  const toolTip = renderer.create(<TooltipIcon id={"hide"} />).toJSON();
  expect(toolTip).toMatchSnapshot();
});

test.each(buttons)(
  "%s tooltip only visible on hover",
  async (button) => {
    render(<TooltipIcon id={button.id} />);
    expect(screen.queryByText(button.text)).not.toBeInTheDocument();
    userEvent.hover(screen.getByTestId(button.symbol));
    expect(screen.getByText(button.text)).toBeInTheDocument();
    userEvent.unhover(screen.getByTestId(button.symbol));
    await waitFor(() => {
      expect(screen.queryByText(button.text)).not.toBeInTheDocument();
    });
  }
);

test("play tooltip only visible on hover", async () => {
  render(<TooltipIcon id="play" paused={true} />);
  expect(screen.queryByText("Play Visualisation")).not.toBeInTheDocument();
  userEvent.hover(screen.getByTestId("PlayArrowIcon"));
  expect(screen.getByText("Play Visualisation")).toBeInTheDocument();
  userEvent.unhover(screen.getByTestId("PlayArrowIcon"));
  await waitFor(() => {
    expect(screen.queryByText("Play Visualisation")).not.toBeInTheDocument();
  });
});

test("pause tooltip only visible on hover", async () => {
  render(<TooltipIcon id="play" paused={false} />);
  expect(screen.queryByText("Pause Visualisation")).not.toBeInTheDocument();
  userEvent.hover(screen.getByTestId("PauseOutlinedIcon"));
  expect(screen.getByText("Pause Visualisation")).toBeInTheDocument();
  userEvent.unhover(screen.getByTestId("PauseOutlinedIcon"));
  await waitFor(() => {
    expect(screen.queryByText("Pause Visualisation")).not.toBeInTheDocument();
  });
});

test("hide graph tooltip only visible on hover", async () => {
  render(<TooltipIcon id="hide" backgroundOn={true} />);
  expect(
    screen.queryByText("Hide original graph edges")
  ).not.toBeInTheDocument();
  userEvent.hover(screen.getByTestId("ToggleOffOutlinedIcon"));
  expect(screen.getByText("Hide original graph edges")).toBeInTheDocument();
  userEvent.unhover(screen.getByTestId("ToggleOffOutlinedIcon"));  
  await waitFor(() => {
    expect(screen.queryByText("Hide original graph edges")).not.toBeInTheDocument();
  });
});

test("show graph tooltip only visible on hover", async () => {
  render(<TooltipIcon id="hide" backgroundOn={false} />);
  expect(
    screen.queryByText("Show original graph edges")
  ).not.toBeInTheDocument();
  userEvent.hover(screen.getByTestId("ToggleOnOutlinedIcon"));
  expect(screen.getByText("Show original graph edges")).toBeInTheDocument();
  userEvent.unhover(screen.getByTestId("ToggleOnOutlinedIcon"));
  await waitFor(() => {
    expect(screen.queryByText("Show original graph edges")).not.toBeInTheDocument();
  });
});

// test("Clicking pause/play toggles correctly", async () => {
//   render(<TooltipIcon id="play" paused={true} />);
//   expect(screen.queryByTestId("PlayArrowIcon")).toBeInTheDocument();
//   expect(screen.queryByTestId("PauseOutlinedIcon")).not.toBeInTheDocument();
//   fireEvent.click(screen.getByTestId("PlayArrowIcon"));
//   await waitFor(() => {
//     expect(screen.queryByTestId("PauseOutlinedIcon")).toBeInTheDocument();
//     expect(screen.queryByTestId("PlayArrowIcon")).not.toBeInTheDocument();
//   });
// });

// test("TooltipIcon component matches snapshot for play", () => {
//   const toolTip = renderer.create(
//     <TooltipIcon
//       symbol={playPause}
//       symbolToolTip="Play Visualisation"
//       id="play"
//     />
//   );
//   expect(toolTip).toMatchSnapshot();
// });
