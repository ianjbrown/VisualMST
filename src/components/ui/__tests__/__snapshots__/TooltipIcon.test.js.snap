// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`back matches snapshot 1`] = `
<svg
  aria-hidden={true}
  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
  data-testid="ArrowBackIosNewOutlinedIcon"
  focusable="false"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  viewBox="0 0 24 24"
>
  <path
    d="M17.77 3.77 16 2 6 12l10 10 1.77-1.77L9.54 12z"
  />
</svg>
`;

exports[`first matches snapshot 1`] = `
<svg
  aria-hidden={true}
  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
  data-testid="FirstPageIcon"
  focusable="false"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  viewBox="0 0 24 24"
>
  <path
    d="M18.41 16.59 13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
  />
</svg>
`;

exports[`forward matches snapshot 1`] = `
<svg
  aria-hidden={true}
  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
  data-testid="ArrowForwardIosOutlinedIcon"
  focusable="false"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  viewBox="0 0 24 24"
>
  <path
    d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
  />
</svg>
`;

exports[`hide matches snapshot 1`] = `
<div
  className="barToggle"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium css-1ddn8qy-MuiSvgIcon-root"
    data-testid="ToggleOnOutlinedIcon"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zm0-7c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
    />
  </svg>
</div>
`;

exports[`last matches snapshot 1`] = `
<svg
  aria-hidden={true}
  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
  data-testid="LastPageIcon"
  focusable="false"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  viewBox="0 0 24 24"
>
  <path
    d="M5.59 7.41 10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
  />
</svg>
`;

exports[`play matches snapshot 1`] = `
<div
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
    data-testid="PauseOutlinedIcon"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"
    />
  </svg>
</div>
`;
